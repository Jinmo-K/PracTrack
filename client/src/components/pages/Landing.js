import React, { Component } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import TextField from '@material-ui/core/TextField';
// Functions
import { loginUser, resetAuthErrors } from "../../actions/authActions";

/** 
 *   Public homepage
 */
class Landing extends Component {
  constructor() {
    super();
    this.state = {
      email: "",
      password: "",
      errors: {}
    };
  }

  componentDidMount() {
    if (this.props.auth.isAuthenticated) {
      this.props.history.push("/");
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.errors) {
      this.setState({
        errors: nextProps.errors
      });
    }
  }

  componentWillUnmount() {
    this.props.resetAuthErrors(['email', 'password']);
  }

  onChange = e => {
    this.setState({ [e.target.id]: e.target.value });
  };

  onFocus = e => {
    this.setState({ 
      errors: {...this.state.errors, [e.target.id]: ''}
    });
  };

  onSubmit = e => {
    e.preventDefault();
    
    const userData = {
      email: this.state.email,
      password: this.state.password
    };

    this.props.loginUser(userData);
  };

  render() {
    const { errors } = this.state;
    return (
      <div className='row justify-content-center login mx-auto'>
        <form className='col-sm-10 col-md-7 col-lg-5 p-2 pt-sm-5' noValidate onSubmit={this.onSubmit}>
          <h2 className='sr-only'>Login form</h2>
          <div className='form-header login-header'>
            <h2>
              <img className="form-logo" src="../../../images/logo-md.png" alt="PracTrack logo"></img> 
              Login to continue
            </h2>
          </div>
          <div className='form-group'>
            <TextField 
              fullWidth
              id='email'
              label='Email'
              variant='outlined'
              error={!!errors.email}
              helperText={errors.email}
              value={this.state.email}
              onFocus={this.onFocus}
              type='email'
              onChange={this.onChange}
            />
          </div>
          <div className='form-group'>
            <TextField 
              fullWidth
              label='Password'
              id="password"
              variant='outlined'
              type='password'
              error={!!errors.password}
              helperText={errors.password}
              value={this.state.password}
              onFocus={this.onFocus}
              onChange={this.onChange}
            />
          </div>
          <div className='form-group'>
            <button className='btn btn-block' type='submit' style={{fontWeight: 500}}>Login</button>
          </div>
          <Link className='forgot' to='/register'>DON'T HAVE AN ACCOUNT? <u>CLICK HERE</u> TO SIGN UP</Link>
        </form>
      </div>
    );
  }
}

Landing.propTypes = {
  /** Auth store state */
  auth: PropTypes.object.isRequired,
  /** Dispatches action to login a user */
  loginUser: PropTypes.func.isRequired,
  /** Dispatches action to reset auth errors */
  resetAuthErrors: PropTypes.func.isRequired,
  /** Errors store state for auth-related errors */
  errors: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { loginUser, resetAuthErrors }
)(Landing);

export { Landing };
