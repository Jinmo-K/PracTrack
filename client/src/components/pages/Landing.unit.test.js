import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router} from 'react-router-dom';
import { Landing } from './Landing';

describe('<Landing />', () => {
  let wrapper;
  let props = {
    history: { push: jest.fn() },
    auth: {
      isAuthenticated: false
    },
    errors: {},
    loginUser: jest.fn(),
    resetAuthErrors: jest.fn()
  }

  beforeEach(() => {
    jest.clearAllMocks();
    props.auth.isAuthenticated = false;
    props.errors = {};
    wrapper = shallow(<Landing {...props} />);
  });

  it('Renders initial layout when user is not logged in', () => {
    let tree = renderer
      .create(
        <Router>
          <Landing {...props} />
        </Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should redirect if user is logged in', () => {
    props.auth.isAuthenticated = true;
    wrapper = shallow(<Landing {...props} />);
    expect(props.history.push.mock.calls[0]).toEqual(['/']);
  });

  it('Should reflect user changes in email field', () => {
    let id = 'email';
    wrapper.find('#'+id).simulate('change', { target: { id, value: 'test' }});
    expect(wrapper.find('#'+id).props().value).toEqual('test');
  });

  it('Should reflect user changes in password field', () => {
    let id = 'password';
    wrapper.find('#'+id).simulate('change', { target: { id, value: 'test' }});
    expect(wrapper.find('#'+id).props().value).toEqual('test');
  });

  it('Should call loginUser when submit button is clicked', () => {
    let mockEvent = {
      preventDefault: jest.fn(),
    };
    wrapper.find('form').simulate('submit', mockEvent);
    expect(mockEvent.preventDefault).toBeCalledTimes(1);
    expect(props.loginUser).toBeCalledTimes(1);
  });

  it('Should render email error', () => {
    wrapper.setState({ errors: { email: 'email error' } });
    expect(wrapper.find('#email').props().helperText).toEqual('email error');
    expect(wrapper.find('#email').props().error).toEqual(true);
  });

  it('Should not render email error', () => {
    wrapper.setState({ errors: { email: '' } });
    expect(wrapper.find('#email').props().error).toEqual(false);
  });

  it('Should render password error', () => {
    wrapper.setState({ errors: { password: 'password error' } });
    expect(wrapper.find('#password').props().helperText).toEqual('password error');
    expect(wrapper.find('#password').props().error).toEqual(true);
  });

  it('Should not render password error', () => {
    wrapper.setState({ errors: { password: '' } });
    expect(wrapper.find('#password').props().error).toEqual(false);
  });

  it('Should reset email error on focus', () => {
    let id = 'email';
    wrapper.setState({ errors: { email: 'email error' } });
    wrapper.find('#'+id).simulate('focus', { target: { id }});
    expect(wrapper.state().errors.email).toEqual('');
  });

  it('Should reset password error on focus', () => {
    let id = 'password';
    wrapper.setState({ errors: { password: 'password error' } });
    wrapper.find('#'+id).simulate('focus', { target: { id }});
    expect(wrapper.state().errors.password).toEqual('');
  });

  it('Should update errors on receiving new props', () => {
    let nextProps = { 
      errors: { 
        email: 'email error', 
        password: 'password error'
      }
    };
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(wrapper.state().errors.email).toEqual('email error');
    expect(wrapper.state().errors.password).toEqual('password error');
  });

  it('Should not update state on new props without errors', () => {
    wrapper.setState({ errors: { email: 'email error', password: 'password error' } });
    let nextProps = {};
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(wrapper.state().errors.email).toEqual('email error');
    expect(wrapper.state().errors.password).toEqual('password error');
  });

  it('Should dispatch resetAuthErrors on unmount', () => {
    wrapper.instance().componentWillUnmount();
    expect(props.resetAuthErrors).toBeCalledTimes(1);
    expect(props.resetAuthErrors.mock.calls[0][0].includes('email')).toEqual(true);
    expect(props.resetAuthErrors.mock.calls[0][0].includes('password')).toEqual(true);
  });
});
