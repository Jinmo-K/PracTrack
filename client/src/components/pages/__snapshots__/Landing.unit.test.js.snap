// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Landing /> Renders initial layout when user is not logged in 1`] = `
<div
  className="row justify-content-center login mx-auto"
>
  <form
    className="col-sm-10 col-md-7 col-lg-5 p-2 pt-sm-5"
    noValidate={true}
    onSubmit={[Function]}
  >
    <h2
      className="sr-only"
    >
      Login form
    </h2>
    <div
      className="form-header login-header"
    >
      <h2>
        <img
          alt="PracTrack logo"
          className="form-logo"
          src="../../../images/logo-md.png"
        />
        Login to continue
      </h2>
    </div>
    <div
      className="form-group"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="email"
          id="email-label"
        >
          Email
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="email"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-61"
            >
              <span>
                Email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-61"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-block"
        style={
          Object {
            "fontWeight": 500,
          }
        }
        type="submit"
      >
        Login
      </button>
    </div>
    <a
      className="forgot"
      href="/register"
      onClick={[Function]}
    >
      DON'T HAVE AN ACCOUNT? 
      <u>
        CLICK HERE
      </u>
       TO SIGN UP
    </a>
  </form>
</div>
`;
